godot-core/src/builtin/basis.rs

```diff
@@ -0,0 +1,980 @@
+        Vector3::new(self[0].x, self[1].x, self[2].x)
+    }
+
+    /// Set the values of the first column of the matrix.
+    pub fn set_col_x(&mut self, col: Vector3) {
+        self[0].x = col.x;
+        self[1].x = col.y;
+        self[2].x = col.z;
+    }
+
+    /// Returns the second column of the matrix,
+    ///
+    /// _Godot equivalent: `Basis.y`_
+    #[must_use]
+    pub fn col_y(&self) -> Vector3 {
+        Vector3::new(self[0].y, self[1].y, self[2].y)
+    }
+
+    /// Set the values of the second column of the matrix.
+    pub fn set_col_y(&mut self, col: Vector3) {
+        self[0].y = col.x;
+        self[1].y = col.y;
+        self[2].y = col.z;
+    }
+
+    /// Returns the third column of the matrix,
+    ///
+    /// _Godot equivalent: `Basis.z`_
+    #[must_use]
+    pub fn col_z(&self) -> Vector3 {
+        Vector3::new(self[0].z, self[1].z, self[2].z)
+    }
+
+    /// Set the values of the third column of the matrix.
+    pub fn set_col_z(&mut self, col: Vector3) {
+        self[0].z = col.x;
+        self[1].z = col.y;
+        self[2].z = col.z;
+    }
+}
+
+impl Display for Basis {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        // Godot output:
+        // [X: (1, 0, 0), Y: (0, 1, 0), Z: (0, 0, 1)]
+        // Where X,Y,Z are the columns
+        let [Vector3 {
+            x: x1,
+            y: x2,
+            z: x3,
+        }, Vector3 {
+            x: y1,
+            y: y2,
+            z: y3,
+        }, Vector3 {
+            x: z1,
+            y: z2,
+            z: z3,
+        }] = self.to_cols();
+
+        f.write_fmt(format_args!(
+            "[X: ({x1}, {x2}, {x3}), Y: ({y1}, {y2}, {y3}), Z: ({z1}, {z2}, {z3})]"
+        ))
+    }
+}
+
+impl From<Quaternion> for Basis {
+    fn from(quat: Quaternion) -> Self {
+        Basis::from_quat(quat)
+    }
+}
+
+impl GlamConv for Basis {
+    type Glam = glam::f32::Mat3;
+}
+
+impl GlamType for glam::f32::Mat3 {
+    type Mapped = Basis;
+
+    fn to_front(&self) -> Self::Mapped {
+        Basis::from_rows_array(&self.to_cols_array()).transposed()
+    }
+
+    fn from_front(mapped: &Self::Mapped) -> Self {
+        Self::from_cols_array(&mapped.to_rows_array()).transpose()
+    }
+}
+
+impl GlamType for glam::f32::Mat3A {
+    type Mapped = Basis;
+
+    fn to_front(&self) -> Self::Mapped {
+        Basis::from_rows_array(&self.to_cols_array()).transposed()
+    }
+
+    fn from_front(mapped: &Self::Mapped) -> Self {
+        Self::from_cols_array(&mapped.to_rows_array()).transpose()
+    }
+}
+
+impl Index<usize> for Basis {
+    type Output = Vector3;
+
+    /// Indexes into the basis by rows.
+    #[inline]
+    fn index(&self, index: usize) -> &Self::Output {
+        &self.rows[index]
+    }
+}
+
+impl IndexMut<usize> for Basis {
+    /// Indexes into the basis by rows.
+    #[inline]
+    fn index_mut(&mut self, index: usize) -> &mut Self::Output {
+        &mut self.rows[index]
+    }
+}
```

Comment:
Same topic as before 😉

I'd wait with indexing traits until we have a clear use case. Maybe we can then find something that works for vectors, transform, basis, and color.

Reply:
So you're suggesting i should add `row_a()` methods etc? Because currently the only two ways to do this is by doing `basis[index]` and `basis.rows[index]`.

Reply:
I would wait with `row_a()`, since the user has public access to the rows. `basis.rows[index]` should be enough for the moment.

But we should probably rethink this later on a level that feels consistent across vectors, matrices, projections, colors etc. For example, instead of index access, we can think about slice access (read-only and mutable). But this is beyond the scope of this PR.